test/Dialect/LLHD/Transforms/hoist-signals.mlir

```diff
@@ -92,10 +92,131 @@ hw.module @DontHoistLocalSignals() {
+    llhd.wait yield (%w : i42), ^bb2
+  ^bb2:
+    // CHECK-NEXT: ^bb2:
+    // CHECK-NEXT: cf.br ^bb3
+    cf.br ^bb3
+  ^bb3:
+    // CHECK-NEXT: ^bb3:
+    // CHECK-NOT: llhd.drv
+    llhd.drv %a, %u after %0 : !hw.inout<i42>
+    llhd.drv %b, %v after %0 : !hw.inout<i42>
+    llhd.drv %c, %u after %1 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt %w, %v, [[EPS]], %false : i42, i42, !llhd.time, i1
+    llhd.halt %w : i42
+  }
+  // CHECK-NEXT: }
+  // CHECK-NEXT: llhd.drv %a, %u after [[DEL]] :
+  // CHECK-NEXT: llhd.drv %b, [[RES]]#1 after [[DEL]] :
+  // CHECK-NEXT: llhd.drv %c, %u after [[RES]]#2 :
+  // CHECK-NEXT: llhd.drv %d, %u after [[DEL]] if [[RES]]#3 :
+  // CHECK-NEXT: call @use_i42([[RES]]#0)
+  func.call @use_i42(%2) : (i42) -> ()
+}
+
+// CHECK-LABEL: @DontHoistDrivesAcrossSideEffects
+hw.module @DontHoistDrivesAcrossSideEffects() {
+  %0 = llhd.constant_time <0ns, 0d, 1e>
+  %c0_i42 = hw.constant 0 : i42
+  %a = llhd.sig %c0_i42 : i42
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: llhd.drv %a
+    llhd.drv %a, %c0_i42 after %0 : !hw.inout<i42>
+    // CHECK-NEXT: call @maybe_side_effecting
+    func.call @maybe_side_effecting() : () -> ()
+    llhd.halt
+  }
+}
+
+// CHECK-LABEL: @DontHoistDrivesOfSlotsWithUnsavoryUsers
+hw.module @DontHoistDrivesOfSlotsWithUnsavoryUsers(inout %c: i42) {
+  %0 = llhd.constant_time <0ns, 0d, 1e>
+  %c0_i42 = hw.constant 0 : i42
+  %a = llhd.sig %c0_i42 : i42
+  %b = llhd.sig %c0_i42 : i42
+  // Unsavory uses outside of process don't affect hoistability.
+  // CHECK: call @use_inout_i42(%a)
+  func.call @use_inout_i42(%a) : (!hw.inout<i42>) -> ()
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NOT: llhd.drv
+    llhd.drv %a, %c0_i42 after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: call @use_inout_i42(%b)
+    func.call @use_inout_i42(%b) : (!hw.inout<i42>) -> ()
+    // CHECK-NEXT: llhd.drv %b
+    llhd.drv %b, %c0_i42 after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: llhd.drv %c
+    llhd.drv %c, %c0_i42 after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+}
+
+// CHECK-LABEL: @OnlyHoistLastDrive
+hw.module @OnlyHoistLastDrive(in %u: i42, in %v: i42) {
+  %0 = llhd.constant_time <0ns, 0d, 1e>
+  %c0_i42 = hw.constant 0 : i42
+  %a = llhd.sig %c0_i42 : i42
+  %b = llhd.sig %c0_i42 : i42
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: llhd.drv %a, %u
+    llhd.drv %a, %u after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.drv %b, %v
+    llhd.drv %b, %v after %0 : !hw.inout<i42>
+    // CHECK-NOT: llhd.drv %a, %v
+    llhd.drv %a, %v after %0 : !hw.inout<i42>
+    // CHECK-NOT: llhd.drv %b, %u
+    llhd.drv %b, %u after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt
     llhd.halt
   }
+  // CHECK: llhd.drv %a, %v
+  // CHECK: llhd.drv %b, %u
```

Comment:
@maerhart noticed an issue here: the hoisted drives can't directly use the module-level values `%u` and `%v`, since any changes to them would cause the updated value to be driven onto `%a` and `%b`. Instead, these values must still be process results, such that they only update once when hitting the `llhd.halt`, and never again. Same for `llhd.wait`.